### PR TITLE
Vector Search - `EXPERIMENTAL`

### PR DESCRIPTION
**Pull request by gmourier**
🤖  [API Diff](https://github.com/meilisearch/specifications/pull/248#issuecomment-1609060852)

---

# Summary

Describes the experimental vector search related properties and capabilities

---

# Changes

- Introduces `vector` and `_semanticScore` properties on `search-api.md`
- Introduces the special properties `_vectors` for a document body
- Describes related errors
    - `invalid_search_vector`
    - `invalid_document_vectors_field`
- Add the experimental properties on open-api.yaml
- Describes added telemetry

# Out Of Scope

N/A

---

# Attention To Reviewers

N/A

---

## Misc

- [x] Update OpenAPI specification file
- [x] Update telemetry datapoints _(if needed; Apply the `Telemetry` label)_

### CODE REVIEW

File: text/0118-search-api.md [hunk 1]
```diff
@@ -944,11 +956,12 @@ A search result can contain special properties. See [3.2.1.1. `hit` Special Prop
...
-|----------------------------------------------|---------|----------|
-| [`_geoDistance`](#32111-geodistance)         | Integer | False    |
-| [`_formatted`](#32112-formatted)             | Object  | False    |
-| [`_matchesPosition`](#32113-matchesposition) | Object  | False    |
+| Field                                                             | Type    | Required |
+|-------------------------------------------------------------------|---------|----------|
+| [`_geoDistance`](#32111-geodistance)                              | Integer | False    |
+| [`_formatted`](#32112-formatted)                                  | Object  | False    |
+| [`_matchesPosition`](#32113-matchesposition)                      | Object  | False    |
+| [`_semanticScore`](#32114-semanticsimilarity) `EXPERIMENTAL` | Float   | False    |
```
Comment (gmourier) [inline text/0118-search-api.md:964]:
```suggestion
| [`_semanticScore`](#32114-semanticscore) `EXPERIMENTAL` | Float   | False    |
```

File: text/0034-telemetry-policies.md [hunk 2]
```diff
@@ -268,6 +269,8 @@ This property allows us to gather essential information to better understand on
 | filter.avg_criteria_number | The average number of filter criteria among all the requests containing the `filter` parameter in the aggregated event. `"filter": []` equals to `0` while not sending `filter` does not influence the average in the aggregated event. | `4` |
 | filter.most_used_syntax | The most used filter syntax among all the requests containing the requests containing the `filter` parameter in the aggregated event. `string` / `array` / `mixed` | `mixed` |
 | q.max_terms_number | The maximum number of terms for the `q` parameter among all requests in the aggregated event. | `5` |
+| vector.max_vector_size | The maximum number of dimensions for the `vector` parameter among all requests in the aggregated event. | `1536` |
+| vector.max_vector_size | The maximum number of dimensions for the `vector` parameter among all requests in the aggregated event. | `1536` |
```
Comment (gmourier) [inline text/0034-telemetry-policies.md:273]:
```suggestion
```

File: text/0034-telemetry-policies.md [hunk 3]
```diff
@@ -122,6 +122,7 @@ The collected data is sent to [Segment](https://segment.com/). Segment is a plat
 | `filter.with_geoBoundingBox`            | `true` if the filter rule `_geoBoundingBox` was used in this batch, otherwise `false`| false | `Documents Searched POST`, `Documents Searched GET` |
 | `filter.most_used_syntax`               | Most used filter syntax among all requests containing the `filter` parameter in this batch | string | `Documents Searched POST`, `Documents Searched GET` |
 | `q.max_terms_number`                    | Highest number of terms given for the `q` parameter in this batch | 5 | `Documents Searched POST`, `Documents Searched GET` |
+| `vector.max_vector_size`                | Highest number of dimensions given for the `vector` parameter in this batch | 1536 | `Documents Searched POST`, `Documents Searched GET` |
```
Comment (gmourier) [inline text/0034-telemetry-policies.md:125]:
```suggestion
| `vector.max_vector_size`                | Highest number of dimensions given for the `vector` parameter in this batch | 1536 | `Documents Searched POST`, `Documents Searched GET`, `Documents Searched by Multi-Search POST` |
```

File: text/0061-error-format-and-definitions.md [hunk 4]
```diff
@@ -1474,6 +1474,29 @@ HTTP Code: `400 Bad Request`
 
 ---
 
+## invalid_search_vector
+
+`Synchronous`
+
+### Context
+
+This error occurs if a value with a different type than `Array of Float` or `null` for `vector` is specified.
```
Comment (gmourier) [inline text/0061-error-format-and-definitions.md:1483]:
```suggestion
This error occurs for the listed reasons:
- if a value with a different type than `Array of Float` or `null` for `vector` is specified.
- if the vector length differs from the documents `_vectors` length.
```

File: text/0124-documents-api.md [hunk 5]
```diff
@@ -261,7 +261,29 @@ A document represented as a JSON object.
...
+
+Refer to the [geo search specification](0059-geo-search.md)
+
+##### 3.1.3.2. `_vectors` `EXPERIMENTAL`
+
+Type: Array of Float | Array[Array of Float]
+Required: False
+
+Holds a vectorized representation of a document. It is possible to send either one or several vectorized representations of the same document.
+
```
Comment (gmourier) [inline text/0124-documents-api.md:285]:
```suggestion

- 🔴 Sending a value with a different type than `Array of Float`, `Array[Array of Float]` or `null` as a value for `_vectors` returns an [invalid_document_vectors_field](0061-error-format-and-definitions.md#invalid_document_vectors_field) error.
- 🔴 Sending a value for `_vectors` whose length differs from another document `_vectors` field returns an [invalid_document_vectors_field](0061-error-format-and-definitions.md#invalid_document_vectors_field) error.
```

File: text/0034-telemetry-policies.md [hunk 6]
```diff
@@ -122,6 +122,7 @@ The collected data is sent to [Segment](https://segment.com/). Segment is a plat
 | `filter.with_geoBoundingBox`            | `true` if the filter rule `_geoBoundingBox` was used in this batch, otherwise `false`| false | `Documents Searched POST`, `Documents Searched GET` |
 | `filter.most_used_syntax`               | Most used filter syntax among all requests containing the `filter` parameter in this batch | string | `Documents Searched POST`, `Documents Searched GET` |
 | `q.max_terms_number`                    | Highest number of terms given for the `q` parameter in this batch | 5 | `Documents Searched POST`, `Documents Searched GET` |
+| `vector.max_vector_size`                | Highest number of dimensions given for the `vector` parameter in this batch | 1536 | `Documents Searched POST`, `Documents Searched GET`, `Documents Searched by Multi-Search POST` |
```
Comment (gmourier) [inline text/0034-telemetry-policies.md:125]: WIP Not working on Amplitude

File: open-api.yaml [hunk 7]
```diff
@@ -1680,6 +1685,8 @@ paths:
 
         > info
         > Use the reserved `_geo` object to add geo coordinates to a document. `_geo` is an object made of `lat` and `lng` field.
+        >
+        > Use the reserved `_vectors` arrays of floats to add embeddings to a document. `_vectors` is an array of floats or multiple arrays of floats in an outter array.
```
Comment (macraig) [inline open-api.yaml:1689]:
```suggestion
        > Use the reserved `_vectors` arrays of floats to add embeddings to a document. `_vectors` is an array of floats or multiple arrays of floats in an outer array.
```

File: open-api.yaml [hunk 8]
```diff
@@ -1734,6 +1741,8 @@ paths:
...
```
Comment (macraig) [inline open-api.yaml:1745]:
```suggestion
        > Use the reserved `_vectors` arrays of floats to add embeddings to a document. `_vectors` is an array of floats or multiple arrays of floats in an outer array.
```

File: text/0061-error-format-and-definitions.md [hunk 9]
```diff
@@ -1474,6 +1474,31 @@ HTTP Code: `400 Bad Request`
...
 ---
 
+## invalid_search_vector
+
+`Synchronous`
+
+### Context
+
+This error occurs for the listed reasons:
+- if a value with a different type than `Array of Float` or `null` for `vector` is specified.
```
Comment (macraig) [inline text/0061-error-format-and-definitions.md:1484]: Is `Array of Float` an existing type?

File: text/0061-error-format-and-definitions.md [hunk 10]
```diff
@@ -1863,6 +1888,38 @@ These errors occurs when the `_geo` field of a document payload is not valid. Ei
 
 ---
 
+## invalid_document_vectors_field
+
+`Asynchronous`
+
+### Context
+
+These errors occurs when the `_vectors` field of a document payload is not valid. Either due to the type of it or the number of dimensions.
```
Comment (macraig) [inline text/0061-error-format-and-definitions.md:1897]:
```suggestion
This error occurs when the `_vectors` field of a document payload is not valid either due to the type of it or the number of dimensions.
```

File: text/0118-search-api.md [hunk 11]
```diff
@@ -912,22 +913,34 @@ The documents containing ALL the query words (i.e. in the `q` parameter) are ret
 
 Only the documents containing ALL the query words (i.e. in the `q` parameter) are returned by Meilisearch. If Meilisearch doesn't have enough documents to fit the requested `limit`, it returns the documents found without trying to match more documents.
 
+#### 3.1.18. `vector` `EXPERIMENTAL`
+
+- Type: Array of Float
+- Required: False
+- Default: []
+
+Request the nearest documents based on the query vector embedding being given.
```
Comment (macraig) [inline text/0118-search-api.md:922]:
```suggestion
Request the nearest documents based on the query vector embedding given.
```

File: text/0034-telemetry-policies.md (hunk 6 above)
Comment (alallema) [reply text/0034-telemetry-policies.md:125]: This also corresponds to the length of the array, doesn't it?

File: text/0061-error-format-and-definitions.md (hunk 9 above)
Comment (bidoubiwa) [reply text/0061-error-format-and-definitions.md:1484]: Is there a point in setting the value of the `vector` field to null?

Comment (gmourier) [reply text/0061-error-format-and-definitions.md:1484]: @alallema 

> An array of arrays and an empty array aren't supposed to work too?

Not at search time; it only accepts an array made of float (no sub-array dimensions) or `null` the default one (equivalent to not sending a vector query)

Comment (gmourier) [reply text/0061-error-format-and-definitions.md:1484]: cc @kero

What is the default value of the `vector` field at search-time? `null` or `[]`.

If we don't accept `[]`, the default value specified should be `null` or the field right?

Comment (alallema) [reply text/0061-error-format-and-definitions.md:1484]: > Is there a point in setting the value of the vector field to null?
Depending on your code if you set a query and want to be sure not to use the vector.
```[2023-07-13T16:05:55Z WARN  meilisearch::search] Ignoring the query string `q` when used with the `vector` parameter.```

File: text/0034-telemetry-policies.md (hunk 6 above)
Comment (maryamsulemani97) [reply text/0034-telemetry-policies.md:125]: @gmourier Is this working now?

**Review by bidoubiwa — APPROVED** [review]
✨

File: text/0034-telemetry-policies.md (hunk 6 above)
Comment (gmourier) [reply text/0034-telemetry-policies.md:125]: cc @macraig (If it's ok from the last Q/A session it can be resolved)

**Pull request merged**